### ISSUE TITLE
Multiframe volume data, non-cubical data domain and why png base64 and not binary array?

### ISSUE THREAD

**Issue by blauwepiet** [maartenbreddels/ipyvolume#97]
I was looking at implementing non-cubical data domains, which kinda worked out with a lot of hacking because it is all hard-coded to be cubical but now all meshes and the domain is scaled based on the limits of the x,y,z domain bounds of the dataset.

But now I was looking at also visualizing volumes together with the meshes but I stumbled on an error while creating the png for the volume (32x32x48). The following error:
```
...
 /home/casperl/Projects/HemoSphere/src/ipyvolume/ipyvolume/serialize.pyc in cube_to_png(grid, vmin, vmax, file)
     81                 warnings.simplefilter("ignore")
     82                 img = PIL.Image.frombuffer("RGBA", (image_width, image_height), data, 'raw')
---> 83                 img.save(file, "png")
     84         return (image_width, image_height), (grid.shape[2], grid.shape[1]), rows, columns, grid.shape[0]
     85 

/home/casperl/anaconda2/envs/hemosphere2/lib/python2.7/site-packages/PIL/Image.pyc in save(self, fp, format, **params)
   1926 
   1927         try:
-> 1928             save_handler(self, fp, filename)
   1929         finally:
   1930             # do what we can to clean up

/home/casperl/anaconda2/envs/hemosphere2/lib/python2.7/site-packages/PIL/PngImagePlugin.pyc in _save(im, fp, filename, chunk, check)
    795 
    796     ImageFile._save(im, _idat(fp, chunk),
--> 797                     [("zip", (0, 0)+im.size, 0, rawmode)])
    798 
    799     chunk(fp, b"IEND", b"")

/home/casperl/anaconda2/envs/hemosphere2/lib/python2.7/site-packages/PIL/ImageFile.pyc in _save(im, fp, tile, bufsize)
    482             if o > 0:
    483                 fp.seek(o, 0)
--> 484             e.setimage(im.im, b)
    485             if e.pushes_fd:
    486                 e.setfd(fp)

SystemError: tile cannot extend outside image
```
And then I also notice that the image width is restricted to 2048, why is that? Why not 4096x4096? This is common max tex size as far as I know. And why is the volume send to the client with png base64 instead of the binary array as the rest of the data? 

Also was wondering why the volume is integrated in the figure widget and not a stand-alone widget? Is that because of legacy coding choices? Because it would look more need if it is also a separate widget like the meshes.

Also I would like to have time dependent volume rendering if that is possible with ipyvolume? It wouldn't be that hard just also implement the sequence index in the figure widget and change the volume texture every sequence frame (already did this once with my own js volume renderer and it worked like a charm back than).

I don't know if the issue board is the best format for these amount of questions and if I have to separate all the questions in different issues but I needed to ask them in order to know your design choices and if I can maybe change them to something else without consequences.

Cheers,
Casper

**Comment by maartenbreddels** [comment]
Hi Casper,

short answer: this is the oldest part of the ipyvolume code. At this time there was no good option to transfer binary data, and actually that is why I started implementing this in ipywidgets. 
I'm not sure why I chose 2048, but this is just a 2d texture, having all 3d slices tiled. 
What you are seeing may be a bug in the tiling since most of the time I've tested it with NxNxN cubes. 

All of these changes really needs to be addressed, have a separate Volume widget, allow binary transfer, possibly animated volume rendering (taking performance into account).
I will be quite busy at least this month, I'm happy to help you out if you want to tackle this, maybe also a videocall may help to explain some things, let me know what I can do to help!

cheers,

Maarten

**Comment by blauwepiet** [comment]
Alright, I thought so already, but good I am happy to help. But for now my code is a bit messy with changes which you don't want in ipyvolume, a lot of hacky code which I still need to refactor to proper solutions haha. 

I'll start with the binary transfer and animation of the volume data and look at the tiling error. But for the separate volume widget (which is only code aesthetics for now) I'll keep it as it is for now, because that looks like quite some work which I don't have the time for now.

**Comment by maartenbreddels** [comment]
Great, you can also do a PR for me te review, even if it is a hack I may give you some useful pointers.

**Comment by blauwepiet** [comment]
Hey Maarten,

I see that you are working on ipyvolume with more frequency again, great!
I also made a lot of changes and improvements to ipyvolume (some duplicate to your latest changes, such as, mesh mem leak fix, dynamic "as square possible" binary volume serialization) in my local repo which I am now setting up for pull requests, but I just wanted to let you know that I am doing this and that we don't do duplicate work.

I'll update this thread with the changes I made if I have everything documented. In the mean time, sorry if this means more double work... I can at least say the following changes I made:
- A separate volume widget to make the figure widget the main widget.
- Dynamic domain shapes, every object added will change the domain shape and it is not only cubical it could also be 10x128x512 for example. (still issue with axis text scaling though)
- Dynamic mesh widget with:
    * dynamic per frame variable vertex and triangle arrays
    * Better handling of meshes, don't make a new one every frame but just change the changed geometry attribute.
    * Reduced the number of materials because depth rendering is now done in depth texture and doesn't have to be a separate material.
- Depth rendering now in textures.
- Dynamic "as square as possible" binary volume serialization, no matter what the shape of the volume it will always try to fit it in a 2d texture as square as possible to efficiently handle irregular volume shapes.
- An interactive geometry slicer widget.
- Modified volume rendering (according to the nvidia gems article with alpha correction) and with different geometry and volume depth interaction (in shader)
- etc (a lot of minor changes and some changes I forgot to mention now)

And the following I am working on right now:
- Better transparency handling for meshes (only meshes not scattering) with depth peeling.

This is a lot of work and changes and I am trying to channel it to you I manageable bite-size chunks. I know not all of the changes I made are going to be in ipyvolume so I try to make things as modular as possible so I can still use those changes in my project.

Hope to have all the changes in separate branches for you so that I can set them as pull requests and eventually merge everything I changed with the current ipyvolume. 

I'll keep you posted!

Casper

**Comment by maartenbreddels** [comment]
Whow 👍 , that is impressive, and indeed, I just pushed some of these changes today and was planning to attack some of them soon, so this is good timing. 
Yes, please try to do them in small parts, also, maybe it is useful to talk over gitter https://gitter.im/maartenbreddels/ipyvolume
Really looking forward to many of these features.

**Comment by maartenbreddels** [comment]
Hi @blauwepiet ,

do you have plans for PR soon, since I'll be attacking some of this now (starting with multi volume rendering).

cheers,

Maarten

**Comment by blauwepiet** [comment]
Hey Maarten,
I have been very busy with something else lately and will have time again after the 24th of April. After that I will continue with the project including this, so the PR can expected after that, if thats ok?

Cheers,
Casper

**Comment by maartenbreddels** [comment]
Of course, no problem. Can you maybe push your changes so I can take a look. Also, what do you think  bites your PR the least? I guess it is better to wait working on multi volume rendering right?

**Comment by blauwepiet** [comment]
Yes, well, I first wanted to refactor it a bit and see which parts could be pushed separately without minimal changes to your source, because there are some parts that I changed which are very drastic which I wanna put in separate widgets so you can decide if you want to change to that widget or not, since it is still your repo ;).

Yes, if the multi volume rendering can wait that would be great. I mean the separate volume from the figure widget in my code almost already allows multi-volumes, except for the blending of the volumes which I didn't look at yet.
Hope I am not causing any inconveniences with this delay though!

**Comment by maartenbreddels** [comment]
ping @blauwepiet 
Could you maybe push your branch to your github repo, that way I can at least take a look, thanks!

**Comment by blauwepiet** [comment]
Hey Maarten,
Finally have time again to work on it with everything else out of the way.
Busy with it as we speak, gonna push my first branches this friday, is that ok? Otherwise I can still push the whole thing though.

**Comment by maartenbreddels** [comment]
Great, looking forward to it!
Would also be great if you can push the whole thing on a separate branch (no PR needed), so I can see which changes are coming.

**Comment by blauwepiet** [comment]
I can't push directly right? or you must give me permission to do so?

**Comment by maartenbreddels** [comment]
But you can just push on your repo. The 'normal' procedure is, you fork my project, you pull that, and make changes to it, and push to your private repo. You can do a PR from a branch, but you don't have to.

**Comment by blauwepiet** [comment]
Yes, then I misunderstood, I thought you wanted me to push to a branch to your repo. But in that case, I have pushed my version here: [https://gitlab.com/caspervanleeuwen/ipyvolume](https://gitlab.com/caspervanleeuwen/ipyvolume)

**Comment by blauwepiet** [comment]
I found that is is a bit more work than I thought it would be and a lot of big functional changes have a lot of minor underlying changes in different files and to put those in different branches costs a lot of time. So bare with me on this one hahaha, it is coming but I can't really give you an estimate.

**Comment by maartenbreddels** [comment]
I cannot see https://gitlab.com/caspervanleeuwen/ipyvolume btw, is it maybe not public?
I guess it also doesn't help that i'm actively developing :)
I'd really like to have the multivolume rendering asap, that is the only thing I want to have before pushing out a 0.5 version, so I really hope your code will help me get there!

**Comment by blauwepiet** [comment]
I didn't notice that it was set to private, i've set it to public now so now you must be able to see it.
